File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int maxSum = Integer.MIN_VALUE;
    private int getSum(TreeNode root) {
        if(root == null) {
            return 0;
        }
        int left = Math.max(getSum(root.left), 0);
        int right = Math.max(getSum(root.right), 0);

        maxSum = Math.max(maxSum, root.val + left + right);
        return root.val + Math.max(left, right);
    }
    public int maxPathSum(TreeNode root) {
 getSum(root);
        return maxSum;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    int Find(TreeNode root, ref int dia)
    {
        if(root is null)
        {
            return 0;
        }
        int left = Find(root.left, ref dia);
        int right =  Find(root.right,ref dia);
        dia = Math.Max(dia, left + right);
        return 1 + Math.Max(left, right);
    }
    public int DiameterOfBinaryTree(TreeNode root) {
        if(root is null) {
            return 0;
        }
        int diameter = 0;
        Find(root, ref diameter);
        return diameter;
    }
}
